File: solutions/assignment1/C++/main.cpp
#include <mysql/mysql.h>
#include <iostream>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <vector>
#include <sstream>

using namespace std;

const char* generateInsertQuery(string &, vector<string> &, vector<string> &);

class MySQLClient {
    MYSQL *conn;
    public:
    
    MySQLClient() {
        this->conn=NULL;
    }

    MySQLClient(MYSQL *conn) {
        this->conn=conn;
    }

    MySQLClient(const char *server, const char *user, const char *password, const char *database) {
        this->conn=mysql_init(NULL);
        if(this->conn == NULL) {
            cerr << "mysql_init() failed" << endl;
            exit(1);
        }

        if(mysql_real_connect(this->conn, server, user, password, database, 0, NULL, 0) == NULL) {
            cerr << "mysql_real_connect() failed, maybe for invalid credentails";
            mysql_close(this->conn);
            exit(1);
        }
    }

    void initConnection(const char *server, const char *user, const char *password, const char *database) {
        this->conn=mysql_init(NULL);
        if(this->conn == NULL) {
            cerr << "mysql_init() failed" << endl;
            exit(1);
        }

        if(mysql_real_connect(this->conn, server, user, password, database, 0, NULL, 0) == NULL) {
            cerr << "mysql_real_connect() failed, maybe for invalid credentails";
            mysql_close(this->conn);
            exit(1);
        }
    }

    void executeQuery(const char* query) { // use for CREATE, DELETE Table don't use for SELECT 
        if (mysql_query(this->conn, query)) {
            cerr << "Some error occured, Error: "<< mysql_error(this->conn) << endl;
            mysql_close(conn);
            exit(1);
        } else {
            cout << "QUERY is successfully executed" << endl;
        }
    }

    void executeSELECTQuery(const char* select_query) {
        if (mysql_query(this->conn, select_query)) {
            cerr << "SELECT failed. Error: " << mysql_error(this->conn) << "\n";
            mysql_close(this->conn);
            exit(1);
        }

        MYSQL_RES *res = mysql_store_result(this->conn);
        if (res == NULL) {
            cerr << "mysql_store_result() failed\n";
            mysql_close(conn);
            exit(1);
        }

        MYSQL_ROW row;
        int num_fields = mysql_num_fields(res);
        while ((row = mysql_fetch_row(res))) {
            for (int i = 0; i < num_fields; i++) {
                cout << (row[i] ? row[i] : "NULL") << " ";
            }
            cout << endl;
        }
        mysql_free_result(res);
    }


    void insertIntoTableFromCSVFile(string &pathName, vector<string> &columns, string &tableName) {
        ifstream in(pathName);
        if(!in.is_open()) {
            cerr << "Error opening files" <<endl;
            exit(1);
        }

        string line;
        getline(in, line);
        while(getline(in, line)) {
            stringstream ss(line);
            vector<string>tokens;
            string token;
            
            while (getline(ss, token, ',')) {
                tokens.push_back(token);
            }

           
            if(tokens.size() != columns.size()) { // Skip Row
                cout << "Row is skipped for invalid data format" << endl;
                continue;
            }

            // const char *insert_query =
            //     "INSERT INTO user (id, first_name, last_name, email, gender, ip_address)"
            //     "VALUES ('Alice', 'Smith', 'alice@example.com', 'Female', '192.168.1.10');";
            
            const char *insert_query = generateInsertQuery(tableName, columns, tokens);
            cout<<insert_query<<endl;
            this->executeQuery(insert_query);
            // cout<<insert_query<<endl;
        }

    }


    void closeConnection() {
        mysql_close(this->conn);
    }


};


const char* generateInsertQuery(string &tableName, vector<string> &columns, vector<string> &tokens) { // be sure that columns.size() == tokens.size()
    static string query;
    query.clear();

    query = "INSERT INTO " + tableName + "(";
    for (size_t i = 0; i < columns.size(); i++) {
        query += columns[i];
        if (i != columns.size() - 1) query += ",";
    }
    query += ") VALUES(";

    for (size_t i = 0; i < tokens.size(); i++) {
        query += "'" + tokens[i] + "'";
        if (i != tokens.size() - 1) query += ",";
    }
    query += ");";

    return query.c_str();
}


int main() {
    const char *server = "192.168.56.1";
    const char *user = "linux-user";
    const char *password = "123";
    const char *database = "testdb";

    MySQLClient *client = new MySQLClient(server, user, password, database);
    
    string pathName="/mnt/c/Users/Jibesh/OneDrive/Desktop/dbms/MOCK_DATA.csv";
    vector<string> columns= {"id", "first_name", "last_name", "email", "gender", "ip_address"};
    string tableName = "user";

    const char *create_table_query = 
        "CREATE TABLE IF NOT EXISTS user ("
        "id INT AUTO_INCREMENT PRIMARY KEY,"
        "first_name VARCHAR(50),"
        "last_name VARCHAR(50),"
        "email VARCHAR(100) UNIQUE,"
        "gender VARCHAR(50),"
        "ip_address VARCHAR(45)"
        ");";

    client->executeQuery("DROP TABLE IF EXISTS user;"); // Delete all row data
    client->executeQuery(create_table_query); // create table
    client->insertIntoTableFromCSVFile(pathName, columns, tableName); // insert all from CSV file
    client->closeConnection();
    
    return 0;
}

File: solutions/assignment1/go/main.go
package main

import (
	"bufio"
	"database/sql"
	"fmt"
	"log"
	"os"
	"strings"

	_ "github.com/go-sql-driver/mysql"
)

// GenerateInsertQuery builds an INSERT query dynamically
func GenerateInsertQuery(tableName string, columns []string, tokens []string) string {
	if len(columns) != len(tokens) {
		return "" // invalid row
	}

	query := fmt.Sprintf("INSERT INTO %s(", tableName)
	query += strings.Join(columns, ",")
	query += ") VALUES("

	values := make([]string, len(tokens))
	for i, token := range tokens {
		// escape single quotes
		values[i] = "'" + strings.ReplaceAll(token, "'", "''") + "'"
	}
	query += strings.Join(values, ",")
	query += ");"

	return query
}

type MySQLClient struct {
	db *sql.DB
}

func NewMySQLClient(server, user, password, database string) *MySQLClient {
	dsn := fmt.Sprintf("%s:%s@tcp(%s)/%s", user, password, server, database)
	db, err := sql.Open("mysql", dsn)
	if err != nil {
		log.Fatalf("sql.Open() failed: %v", err)
	}
	if err := db.Ping(); err != nil {
		log.Fatalf("mysql_real_connect() failed: %v", err)
	}
	return &MySQLClient{db: db}
}

func (c *MySQLClient) ExecuteQuery(query string) {
	_, err := c.db.Exec(query)
	if err != nil {
		log.Fatalf("Query failed: %v\nQuery: %s", err, query)
	} else {
		fmt.Println("QUERY executed successfully")
	}
}

func (c *MySQLClient) ExecuteSelectQuery(query string) {
	rows, err := c.db.Query(query)
	if err != nil {
		log.Fatalf("SELECT failed: %v", err)
	}
	defer rows.Close()

	cols, _ := rows.Columns()
	values := make([]sql.RawBytes, len(cols))
	scanArgs := make([]interface{}, len(values))
	for i := range values {
		scanArgs[i] = &values[i]
	}

	for rows.Next() {
		err = rows.Scan(scanArgs...)
		if err != nil {
			log.Fatal(err)
		}
		for _, col := range values {
			if col == nil {
				fmt.Print("NULL ")
			} else {
				fmt.Print(string(col) + " ")
			}
		}
		fmt.Println()
	}
}

func (c *MySQLClient) InsertIntoTableFromCSVFile(pathName string, columns []string, tableName string) {
	file, err := os.Open(pathName)
	if err != nil {
		log.Fatalf("Error opening file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// skip header
	if scanner.Scan() {
		// ignore first line
	}

	for scanner.Scan() {
		line := scanner.Text()
		tokens := strings.Split(line, ",")

		if len(tokens) != len(columns) {
			fmt.Println("Row skipped due to invalid format:", line)
			continue
		}

		insertQuery := GenerateInsertQuery(tableName, columns, tokens)
		if insertQuery != "" {
			fmt.Println(insertQuery)
			c.ExecuteQuery(insertQuery)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %v", err)
	}
}

func (c *MySQLClient) Close() {
	c.db.Close()
}

func main() {
	server := "192.168.56.1:3306"
	user := "linux-user"
	password := "123"
	database := "testdb"

	client := NewMySQLClient(server, user, password, database)

	tableName := "user"
	columns := []string{"id", "first_name", "last_name", "email", "gender", "ip_address"}
	pathName := "/mnt/c/Users/Jibesh/OneDrive/Desktop/dbms/MOCK_DATA.csv"

	createTableQuery := `
	CREATE TABLE IF NOT EXISTS user (
		id INT AUTO_INCREMENT PRIMARY KEY,
		first_name VARCHAR(50),
		last_name VARCHAR(50),
		email VARCHAR(100) UNIQUE,
		gender VARCHAR(50),
		ip_address VARCHAR(45)
	);`

	client.ExecuteQuery("DROP TABLE IF EXISTS user;")
	client.ExecuteQuery(createTableQuery)
	client.InsertIntoTableFromCSVFile(pathName, columns, tableName)
	client.Close()
}
